frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -647,15 +647,63 @@ func dispatchEnv(d *dispatchState, c *instructions.EnvCommand) error {
 }
 
 func dispatchRun(d *dispatchState, c *instructions.RunCommand, proxy *llb.ProxyEnv, sources []*dispatchState, dopt dispatchOpt) error {
+	var opt []llb.RunOption
+
 	var args []string = c.CmdLine
+	if len(c.Files) > 0 {
+		if len(args) != 1 {
+			return fmt.Errorf("parsing produced an invalid run command: %v", args)
+		}
+
+		if heredoc := parser.MustParseHeredoc(args[0]); heredoc != nil {
+			if d.image.OS != "windows" && strings.HasPrefix(c.Files[0].Data, "#!") {
+				// This is a single heredoc with a shebang, so create a file
+				// and run it.
+				// NOTE: choosing to expand doesn't really make sense here, so
+				// we silently ignore that option if it was provided.
+				sourcePath := "/"
+				destPath := "/dev/pipes/"
+
+				f := c.Files[0].Name
+				data := c.Files[0].Data
+				if c.Files[0].Chomp {
+					data = parser.ChompHeredocContent(data)
+				}
+				st := llb.Scratch().Dir(sourcePath).File(llb.Mkfile(f, 0755, []byte(data)))
+
+				mount := llb.AddMount(destPath, st, llb.SourcePath(sourcePath), llb.Readonly)
+				opt = append(opt, mount)
+
+				args[0] = path.Join(destPath, f)
```

Comment:
Hm, I've just realized a slightly interesting case. We *probably* don't want to prepend the shell if we're creating and running a separate file, especially since now we're not relying on the shell's behavior if the exec call fails.

To give an example:
```dockerfile
SHELL ["python", "-c"]
RUN <<EOF
print("this runs under python")
EOF

RUN <<EOF
#!/usr/bin/env python
print("this doesn't work, because python can't run files!")
EOF
```
Dropping the shell prepend for this branch allows both cases to properly work, and I can't *think* of a case where we'd rely on which `SHELL` was set for when we have a shebang... This is kind of a continuation of the fake-heredoc support.

Reply:
I don't think this is a very important case. It makes sense why it doesn't work and easy for user to do the right thing. I consider `SHELL ["python", "-c"]` anti-pattern anyway.

Reply:
Yup, it's been a while looking/thinking at this feature, I think I'm probably seeing use cases where there aren't any :smile: It's definitely an edge case and an anti-pattern.

That said, it's a super small fix (we just force `PrependShell` to `false`), and is an extension to fake heredocs, and as far as I can tell doesn't haven't any downsides, aside from a tiny bit of added complexity.

Up to you, I can have a commit ready soon, or we can drop it entirely.